fix(pricing_rule): transaction rule doesn't got apply on frontend

pipech opened the pull request:
## Fix for > [Bug] V12 - Pricing rule: Transaction rule doesn't got apply on frontend #22159

Apply transaction rule on frontend through `transaction.js` and `point_of_sale.js`.

This will work on both Sales Invoice page and Online POS page and both product discount and price discount.

pipech committed:
fix(pricing_rule): transaction discount apply on frontend

rohitwaghchaure commented on erpnext/public/js/controllers/transaction.js:
```diff
@@ -569,6 +569,9 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 									}
+								},
+								() => {
+									me.frm.trigger('qty', cdt, cdn);
```
Why trigger qty?

pipech replied:
To trigger apply_pricing_rule, actually i should change this to `this.apply_pricing_rule(item, true);`

pipech replied:
https://github.com/frappe/erpnext/blob/bc4737a1874b1c021c98fb50ae0c4b2dfecb5510/erpnext/public/js/controllers/transaction.js#L573-L575

rohitwaghchaure commented on erpnext/public/js/controllers/transaction.js:
```diff
@@ -1264,6 +1267,43 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 			}
 		});
+
+		const transaction_pricing_rule = this.frm.call({
```
This should be call in the callback of apply_pricing_rule

pipech replied:
May i ask why?

rohitwaghchaure commented on erpnext/public/js/controllers/transaction.js:
```diff
@@ -1264,6 +1267,43 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
+			callback: function(r) {
+				if (!r.exc && r.message) {
+					r = r.message;
+					if ('price' in r) {
```
why not checking the value of the field price_or_product_discount

pipech replied:
Return doesn't have `apply_transaction_pricing_rule` field.

pipech committed:
fix(pricing_rule): make function more explicit

pipech commented:
@rohitwaghchaure sorry to tag you but could you please review this ?

pipech committed:
fix(pos): fix transaction discount doesn't update

add triggered event when apply_discount_on got change
because sometime callback, come back after cart update taxes value,
so we need to triggred update cart manually

pipech committed:
fix(pos): use discount_amount instead of apply_discount_on

since discount_amount got triggered less often

sagarvora commented on erpnext/accounts/doctype/pricing_rule/pricing_rule.py:
```diff
@@ -175,6 +175,51 @@ def apply_pricing_rule(args, doc=None):
+	# converting dict into doc
+	if isinstance(doc, string_types):
+		doc = json.loads(doc)
+	if doc:
```
I don't know if this condition is needed at all. And if it _is_ needed, what happens if it's falsy?

rohitwaghchaure commented:
@pipech I was doing the testing for the POS, and I have found one issue that while removing the main item from the cart the free items is still remained in the cart. Please check below GIF

![pos_testing_issue](https://user-images.githubusercontent.com/8780500/104319691-39ee9080-5507-11eb-8d33-f38e3b49627a.gif)

I am closing this PR, requesting you to kindly fix this issue and reopen the PR

federicocalvo commented:
@rohitwaghchaure @sagarvora Hi there, this PR is the only one that comes close to a correction in order to be able to use **pricing rules** with the **transaction** condition.

As of today, it still does not work and this PR was not reopened or revised.

Can it at least be reopened? We might be able to see about continuing it with our team.

Thanks

sagarvora commented:
👋🏼 @federicocalvo 

Can you please open another one on `develop`? Or do you intend to use this on the `v12` POS?

federicocalvo commented:
Hi @sagarvora, yes of course, but in our case it would be excluding the POS (I think it is very complex) but in the transactions as Sale Invoice (it should refer to Sales Order and Cuotation as well).

Then someone could separately try to solve it in point_of_sale.js.

sagarvora commented:
> Hi @sagarvora, yes of course, but in our case it would be excluding the POS (I think it is very complex) but in the transactions as Sale Invoice (it should refer to Sales Order and Cuotation as well).
> 
> Then someone could separately try to solve it in point_of_sale.js.

Great, tag / assign me. I'll review it.
